### PR TITLE
Cleanup table_slice API

### PR DESCRIPTION
**Pull request by dominiklohmann**
###  :notebook_with_decorative_cover: Description

This is the first PR to be merged into #1140, doing some much-needed cleanup work on the existing `table_slice` API that will make future changes easier.

###  :memo: Checklist

- [x] All user-facing changes have changelog entries.
- [x] The changes are reflected on [docs.tenzir.com/vast](https://docs.tenzir.com/vast), if necessary.
- [x] The PR description contains instructions for the reviewer, if necessary.

### :dart: Review Instructions

Commit-by-commit is best.

### CODE REVIEW

File: libvast/vast/table_slice_column.hpp
Comment (tobim) [inline libvast/vast/table_slice_column.hpp:42]: I think both of those constructors should be `private`. Instead, we should provide
```
static std::optional<table_slice_column>(table_slice_ptr slice, size_t column);
static std::optional<table_slice_column>(table_slice_ptr slice, std::string_view column);
```
Optional or pointer semantics should not creep into the actual `table_slice_column`.

File: libvast/src/system/explorer.cpp
Comment (tobim) [inline libvast/src/system/explorer.cpp:138]: This exemplifies my issue with the "nullability" of `table_slice_column`. With the way it is defined right now, the `optional` is redundant.

File: libvast/vast/table_slice_column.hpp
Comment (tobim) [inline libvast/vast/table_slice_column.hpp:56]: To me `rows()` is semantically clearer.

File: libvast/src/system/explorer.cpp
Comment (tobim) [inline libvast/src/system/explorer.cpp:152]: This is a strange way to test for validity. Imho value types must be valid by construction.

**Review by tobim — CHANGES_REQUESTED** [review]

File: libvast/vast/table_slice_row.hpp
Comment (tobim) [inline libvast/vast/table_slice_row.hpp:28]: What is the purpose of this class? I think it would be preferable to use `record_view` instead.

File: libvast/vast/table_slice.hpp
Comment (tobim) [inline libvast/vast/table_slice.hpp]: This is very good.

File: libvast/vast/table_slice_row.hpp
Comment (dominiklohmann) [reply libvast/vast/table_slice_row.hpp:28]: We have a story for a future expansion that makes me want to keep this as is.

File: libvast/vast/table_slice_column.hpp
Comment (dominiklohmann) [reply libvast/vast/table_slice_column.hpp:56]: Using `size()` and `index()` allows for using `table_slice_row` and `table_slice_column` interchangeably.

File: libvast/vast/table_slice_column.hpp
Comment (dominiklohmann) [reply libvast/vast/table_slice_column.hpp:42]: I'd argue this only really makes sense for the latter constructor. Will change for that one.

File: libvast/vast/table_slice_row.hpp
Comment (tobim) [reply libvast/vast/table_slice_row.hpp:28]: Can you be more specific? What functionality should be based on this?

Comment (dominiklohmann) [reply libvast/vast/table_slice_row.hpp:28]: Row-major and column-major iteration can be built on top of these.

Comment (tobim) [reply libvast/vast/table_slice_row.hpp:28]: this type would only be useful for row-major iteration, but that iterator should point to `record_view`s.

Comment (dominiklohmann) [reply libvast/vast/table_slice_row.hpp:28]: That would require decoding the whole row beforehand, which we don't necessarily want to do. The data access should be lazy through the encoding, regardless of whether it is MessagePack or Arrow.

Comment (mavam) [reply libvast/vast/table_slice_row.hpp:28]: Clearly, we don't have the use case fleshed out yet. Until then, I propose we put these utilities in the backlog and add them only when needed. Otherwise we carry more complexity with us that we need.

Comment (tobim) [reply libvast/vast/table_slice_row.hpp:28]: We agreed to keep the type around in case we really do need it later. It is much more discoverable inside the tree than somewhere in an old branch.

File: libvast/vast/table_slice_column.hpp
Comment (tobim) [reply libvast/vast/table_slice_column.hpp:43]: As long as we keep the precondition it is ok.

**Review by tobim — APPROVED** [review]

**Comment by dominiklohmann** [comment]
I rebased after the approval to get rid of the autosquash commits. I'll merge into the epic branch once the CI is green.

**Review by tobim — APPROVED** [review]
Tested locally, the changes work. I did not do a performance analysis however.

**Comment by dominiklohmann** [comment]
Rebased one further time after fixing the remaining stack-use-after-scope issues w.r.t `table_slice::layout()` returning by value.

**Pull request merged**
